app/controllers/admin/deploy_groups_controller.rb

```diff
@@ -76,56 +76,89 @@ def deploy_all
   end
 
   def create_all_stages_preview
-    @preexisting_stages, @missing_stages = stages_for_creation
+    @preexisting_stages, @missing_stages = self.class.stages_for_creation(deploy_group)
   end
 
   # No more than one stage, per project, per deploy_group
   # Note: you can call this multiple times, and it will create missing stages, but no redundant stages.
   def create_all_stages
-    _, missing_stages = stages_for_creation
-    missing_stages.each do |template_stage|
-      create_stage_with_group(template_stage)
+    stages_created = self.class.create_all_stages(deploy_group)
+
+    flash[:success] = "Created #{stages_created.length} Stages"
+
+    redirect_to [:admin, deploy_group]
+  end
+
+  def merge_all_stages
+    cloned_stages = deploy_group.stages.where.not(template_stage_id: nil)
+    cloned_stages.each do |stage|
+      merge_stage(stage)
     end
 
     redirect_to [:admin, deploy_group]
   end
 
+  def self.create_all_stages(deploy_group)
+    _, missing_stages = stages_for_creation(deploy_group)
+    missing_stages.map do |template_stage|
+      create_stage_with_group(template_stage, deploy_group)
+    end
+  end
+
   private
 
-  # returns a list of stages already created and list of stages to create (through their template stages)
-  def stages_for_creation
-    environment = deploy_group.environment
-    template_stages = environment.template_stages.all
-    deploy_group_stages = deploy_group.stages.all
-
-    preexisting_stages = []
-    missing_stages = []
-    Project.where(include_new_deploy_groups: true).each do |project|
-      template_stage = template_stages.detect { |ts| ts.project_id == project.id }
-      deploy_group_stage = deploy_group_stages.detect { |dgs| dgs.project.id == project.id }
-      if deploy_group_stage
-        preexisting_stages << deploy_group_stage
-      elsif template_stage
-        missing_stages << template_stage
-      end
-    end
+  def merge_stage(stage)
+    template_stage = stage.template_stage
+
+    return unless template_stage
+    return if template_stage.deploy_groups.include?(stage.deploy_groups.first)
+    return if stage.is_template
```

Comment:
This is just a sanity check. But if it has been turned into a template, other stages could reference it, want to merge into it (for some reason).